### PR TITLE
DEV-2214: Add ColorPicker component

### PR DESCRIPTION
**Pull request by nrh-cklimas**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

n/a

#### What's this PR do?

- Adds a new base component, ColorPicker, which replaces the existing one.
- Refactors the colors section of the style editor modal into a separate TypeScript component.

#### Why are we doing this? How does it help us?

Improves UX and addresses CSP problems with react-color.

#### How should this be manually tested? Please include detailed step-by-step instructions.

I noticed a bug where if you make changes to a style under the Customize section then re-edit it immediately, the change doesn't appear. But reloading the page seems to fix it. The bug seems present on main.

1. Edit a style.
2. Edit colors by clicking the swatch button beside the text field. Confirm that...
   a. Changing the color there changes the color value in the text field beside it.
   b. The icon of the swatch updates based on the brightness of the color (e.g. white on dark colors, black on light colors).
4. Try editing the color text fields and confirm that...
   a. You can't type non hex digits (e.g. the word "no").
   b. If you type a partial hex triplet (like `#1`) then the field value reverts to what it was before you started editing it.
   c. You can copy and paste into the fields freely.
   d. You can type up to 7 characters into each field.
   e. Edits in the text field cause the swatch button to change.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

Yes. The color pickers have changed appearance.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

Updated documentation I think.

#### Are there any smells or added technical debt to note?

See note about the bug in the QA section.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2214

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.

### CODE REVIEW

**Review by GuiMend — COMMENTED** [review]
Color picker working really nicely!

Have just a few comments

File: spa/src/components/base/ColorPicker/ColorPicker.test.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      tree({}, true);
+
+      // These changes happen in the textbox because it appears jsdom normalizes
+      // invalid color values on a color input. That is, triggering a change
+      // event with value 'no' gets normalized to '#000000'.
+
+      fireEvent.change(screen.getByRole('textbox'), { target: { value: '#fff123' } });
+      expect(screen.getByTestId('swatch-input')).toHaveValue('#fff123');
+      fireEvent.change(screen.getByRole('textbox'), { target: { value: 'no' } });
+      expect(screen.getByTestId('swatch-input')).toHaveValue('#fff123');
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/components/base/ColorPicker/ColorPicker.test.tsx:60]: The idea here is to try to put an invalid color value into the component, and we can only do it by making changes in the text box. The comment was to try to explain why we're using the textbox to test the color input.

File: spa/src/components/stylesEditor/StylesEditor.js [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import { useState, useEffect } from 'react';
-import { HelpText, Label } from 'elements/inputs/BaseField.styled';
+import { Label } from 'elements/inputs/BaseField.styled';
 import * as S from './StylesEditor.styled';
 
+// TODO: Pull color section into separate TSX component, test that
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/components/stylesEditor/StylesEditor.js]: Yep, removed it in a9093fb983742bb930db2ef3f49486da23064371

File: spa/src/components/base/ColorPicker/ColorPicker.test.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    it('reverts the value when the field loses focus if the current value is not valid', () => {
+      const onChange = jest.fn();
+
+      tree({ onChange }, true);
+      fireEvent.change(screen.getByRole('textbox'), { target: { value: '#fff123' } });
+      fireEvent.change(screen.getByRole('textbox'), { target: { value: '#111' } });
+      onChange.mockClear();
+      fireEvent.blur(screen.getByRole('textbox'));
+      expect(onChange.mock.calls).toEqual([['#fff123']]);
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/components/base/ColorPicker/ColorPicker.test.tsx]: That makes sense to me. Updated in a9093fb983742bb930db2ef3f49486da23064371

File: spa/src/utilities/isHexColor.test.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { isHexColor } from './isHexColor';
+
+describe('isHexColor', () => {
+  it('returns true for a full hex triplet prefixed by #', () => {
+    expect(isHexColor('#12ab3f')).toBe(true);
+    expect(isHexColor('#12AB3F')).toBe(true);
+    expect(isHexColor('#111111')).toBe(true);
+    expect(isHexColor('#ffffff')).toBe(true);
+  });
+  it('returns false for a hex triplet shorthand', () => expect(isHexColor('#111')).toBe(false));
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/utilities/isHexColor.test.ts:10]: A limitation of native color inputs is they only accept hex triplets, not even shorthands. And they don't do alpha, either. I added a comment in a9093fb983742bb930db2ef3f49486da23064371 to `isHexColor()` to make that clearer.

File: spa/src/utilities/isHexColor.test.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  it('returns false if there are too many characters', () => expect(isHexColor('#12ab3f0')).toBe(false));
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/utilities/isHexColor.test.ts:11]: Same reason as why it doesn't accept shorthandds.

File: spa/src/components/stylesEditor/ColorsEditor.test.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    expect(mockPreview).toBeInTheDocument();
+    expect(mockPreview.dataset.headerColor).toBe(mockColors.cstm_mainHeader);
+    expect(mockPreview.dataset.background).toBe(mockColors.cstm_mainBackground);
+    expect(mockPreview.dataset.panelBackground).toBe(mockColors.cstm_formPanelBackground);
+    expect(mockPreview.dataset.buttons).toBe(mockColors.cstm_CTAs);
+    expect(mockPreview.dataset.accents).toBe(mockColors.cstm_ornaments);
+    expect(mockPreview.dataset.inputBackground).toBe(mockColors.cstm_inputBackground);
+    expect(mockPreview.dataset.inputBorder).toBe(mockColors.cstm_inputBorder);
+  });
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/components/stylesEditor/ColorsEditor.test.tsx:59]: Idea behind this test is to prove ColorsEditor is giving ColorPickerPreview the correct values.

**Review by GuiMend — APPROVED** [review]

**Review by x110dc — APPROVED** [review]

**Pull request merged**